klippy/webhooks.py

```diff
@@ -280,3 +280,136 @@ def get_status(self, eventtime=0.):
+                                        if k in req_items and not callable(v)}
+        else:
+            result = {"status": "Klippy Not Ready"}
+        return result
+
+    def _handle_object_request(self, web_request):
+        if web_request.get_method() != 'GET':
+            raise web_request.error("Invalid Request Method")
+        web_request.send(dict(self.available_objects))
+
+    def _handle_status_request(self, web_request):
+        if web_request.get_method() != 'GET':
+            raise web_request.error("Invalid Request Method")
+        args = web_request.get_args()
+        result = self._process_status_request(args)
+        web_request.send(result)
+
+    def _handle_subscription_request(self, web_request):
+        method = web_request.get_method()
+        if method == 'POST':
+            # add a subscription
+            args = web_request.get_args()
+            if args:
+                self.add_subscripton(args)
+            else:
+                raise web_request.error("Invalid argument")
+        else:
+            # get subscription info
+            result = dict(self.subscriptions)
+            web_request.send(result)
+
+    def add_subscripton(self, new_sub):
+        if not new_sub:
+            return
+        for obj_name, req_items in new_sub.items():
+            if obj_name not in self.available_objects or \
+                    obj_name.startswith("gcode_macro"):
+                logging.info(
+                    "webhooks: Object {%s} not available for subscription"
+                    % (obj_name))
+                continue
+            # validate requested items
+            if req_items:
+                avail_items = set(self.available_objects[obj_name])
+                invalid_items = set(req_items) - avail_items
+                if invalid_items:
+                    logging.info(
+                        "webhooks: Removed invalid items [%s] from "
+                        "subscription request %s" %
+                        (", ".join(invalid_items), obj_name))
+                    req_items = list(set(req_items) - invalid_items)
+                    if not req_items:
+                        # No valid items remaining
+                        continue
+            # Add or update subscription
+            existing_items = self.subscriptions.get(obj_name, None)
+            if existing_items is not None:
+                if req_items == [] or existing_items == []:
+                    # Subscribe to all items
+                    self.subscriptions[obj_name] = []
+                else:
+                    req_items = list(set(req_items) | set(existing_items))
+                    self.subscriptions[obj_name] = req_items
+            else:
+                self.subscriptions[obj_name] = req_items
+        waketime = self.reactor.monotonic() + SUBSCRIPTION_REFRESH_TIME
+        self.reactor.update_timer(self.subscription_timer, waketime)
```

Comment:
Just out of curiosity, why push back the timer at this point?  Why not just let the timer run at its normal time?

Reply:
I start the timer here if it hasn't already been started.  I think its a good idea to add a check and only update the timer if it isn't running.